### PR TITLE
[SPARK-22654][TESTS] Retry Spark tarball download if failed in HiveExternalCatalogVersionsSuite

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

Adds a simple loop to retry download of Spark tarballs from different mirrors if the download fails.

## How was this patch tested?

Existing tests

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala [hunk 1]
```diff
@@ -50,14 +52,24 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
...
 
-  private def downloadSpark(version: String): Unit = {
-    import scala.sys.process._
+  private def tryDownloadSpark(version: String, path: String): Unit = {
+    // Try mirrors a few times until one succeeds
+    for (i <- 0 until 3) {
+      val preferredMirror =
+        Seq("wget", "https://www.apache.org/dyn/closer.lua?preferred=true", "-q", "-O", "-").!!.trim
+      val url = s"$preferredMirror/spark/spark-$version/spark-$version-bin-hadoop2.7.tgz"
+      logInfo(s"Downloading Spark $version from $url")
```
Comment (felixcheung) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: FYI not sure which is the better way, in SparkR we get a list with `http://www.apache.org/dyn/closer.cgi&as_json=1`

```
{
       "backup": [ "http://www-eu.apache.org/dist/", "http://www-us.apache.org/dist/" ],
         "cca2": "us",
          "ftp": [ "ftp://apache.cs.utah.edu/apache.org/", "ftp://apache.mirrors.tds.net/pub/apache.org/", "ftp://ftp.osuosl.org/pub/apache/", "ftp://mirror.reverse.net/pub/apache/" ],
         "http": [ "http://apache.claz.org/", "http://apache.cs.utah.edu/", "http://apache.mesi.com.ar/", "http://apache.mirrors.hoobly.com/", "http://apache.mirrors.ionfish.org/", "http://apache.mirrors.lucidnetworks.net/", "http://apache.mirrors.pair.com/", "http://apache.mirrors.tds.net/", "http://apache.osuosl.org/", "http://download.nextag.com/apache/", "http://ftp.wayne.edu/apache/", "http://mirror.cc.columbia.edu/pub/software/apache/", "http://mirror.cogentco.com/pub/apache/", "http://mirror.jax.hugeserver.com/apache/", "http://mirror.metrocast.net/apache/", "http://mirror.olnevhost.net/pub/apache/", "http://mirror.reverse.net/pub/apache/", "http://mirror.stjschools.org/public/apache/", "http://mirrors.advancedhosters.com/apache/", "http://mirrors.gigenet.com/apache/", "http://mirrors.koehn.com/apache/", "http://mirrors.ocf.berkeley.edu/apache/", "http://mirrors.sonic.net/apache/", "http://supergsego.com/apache/", "http://www.gtlib.gatech.edu/pub/apache/", "http://www.namesdir.com/mirrors/apache/", "http://www.trieuvan.com/apache/" ],
    "path_info": "",
    "preferred": "http://mirror.cc.columbia.edu/pub/software/apache/"
}
```

Comment (felixcheung) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: it is possible to also pass `path` but from testing, it doesn't seem like the path/path_info value is validated in any way.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala [hunk 2]
```diff
@@ -50,14 +52,24 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
...
+      if (Seq("wget", url, "-q", "-P", path).! == 0) {
+        return
+      }
+      logWarning(s"Failed to download Spark $version from $url")
+    }
+    fail(s"Unable to download Spark $version")
```
Comment (felixcheung) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: btw, I've also seen a mirror abruptly ending a download but not getting reported as an error, resulting in an incomplete/corrupted tgz.

it's possible the mirror misreports the response byte size in that case.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.2.

**Pull request closed without merging**